### PR TITLE
bpo-39207: Spawn workers on demand in ProcessPoolExecutor

### PR DESCRIPTION
**Pull request by aeros**
Roughly based on https://github.com/python/cpython/commit/904e34d4e6b6007986dcc585d5c553ee8ae06f95, but with a few substantial differences.

/cc @pitrou @brianquinlan 

https://bugs.python.org/issue39207

Automerge-Triggered-By: @pitrou

### CODE REVIEW

File: Lib/concurrent/futures/process.py [hunk 1]
```diff
@@ -249,6 +252,8 @@ def _process_worker(call_queue, result_queue, initializer, initargs):
         # open files or shared memory that is not needed anymore
         del call_item
 
+        # increment idle process count after worker finishes job
+        idle_worker_semaphore.release()
```
Comment (aeros) [inline Lib/concurrent/futures/process.py]: I considered having this in `_ExecutorManagerThread.run()` at first (both technically work), but I think having it in `_process_worker` instead will result in more immediate reuse of the workers, which helps to further minimize idle CPU time.

**Comment by aeros** [comment]
I think the MacOS failure can be resolved by setting `_idle_workers_semaphore = None` in `executor.shutdown()`, it appears to be an issue with running out of file descriptors. I'll see if that addresses the problem.

**Comment by aeros** [comment]
Hmm... I could potentially try using a weakref of the executor to access the semaphore (and then deleting it) instead of directly using it in `_process_worker`; in theory, this should allow it to be GC'd sooner. For example:

```py
executor = executor_reference()
if executor is not None:
    executor._idle_worker_semaphore.release
del executor
```

But, I'd also be willing to try any other possible solutions to address the MacOS failure. See the [log](https://github.com/python/cpython/pull/19453/checks?check_run_id=575730811#step:6:562) for details.
```

File: Lib/concurrent/futures/process.py (hunk 1 above)
Comment (aeros) [reply Lib/concurrent/futures/process.py]: I ended up needing to release the semaphore in `_ExecutorManagerThread.run()`, see comments below for more information.

**Comment by aeros** [comment]
Moving the semaphore to be accessed through a weakref to the executor addressed the problem. I'm certain it was an issue with the file descriptors, but not 100% sure if it was related to it using an excessive number of them being used at once or if it was trying to access a file descriptor that had already been GC'd.

Ultimately, it results in a slight delay between when the worker is finished to when the idle semaphore is released (compared to releasing it immediately at the end of `_process_worker()`), but it will still be a significant overall improvement.

File: Lib/test/test_concurrent_futures.py [hunk 2]
```diff
@@ -964,6 +970,37 @@ def test_ressources_gced_in_workers(self):
...
+        self.assertEqual(len(executor._processes), 1)
+        executor.shutdown()
+
+    def test_idle_process_reuse_multiple(self):
+        executor = self.executor_type()
+        executor.submit(mul, 12, 7).result()
+        executor.submit(mul, 33, 25)
+        executor.submit(mul, 25, 26).result()
+        executor.submit(mul, 18, 29)
+        self.assertEqual(len(executor._processes), 2)
```
Comment (aeros) [inline Lib/test/test_concurrent_futures.py]: This test might be subject to race conditions, as indicated in https://buildbot.python.org/all/#/builders/296/builds/46:

```
======================================================================
FAIL: test_idle_process_reuse_multiple (test.test_concurrent_futures.ProcessPoolSpawnProcessPoolExecutorTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/root/buildarea/pull_request.angelico-debian-amd64/build/Lib/test/test_concurrent_futures.py", line 1002, in test_idle_process_reuse_multiple
    self.assertEqual(len(executor._processes), 2)
AssertionError: 1 != 2
```

 Instead, I think it would probably make more sense to check `self.assertTrue(len(executor._processes) <= 2)`, after explicitly setting max workers to a specific amount, e.g. `executor = self.executor_type(4)`, and then submitting `_max_workers` jobs. It's perfectly okay if there are less than 2 workers as a result of the jobs being completed quickly, but there should never be more than two if the idle workers are being properly reused (since we directly waited for the `result()` on two of them).

I'll take a look at this again tomorrow, after the buildbot tests have completed.

**Comment by pitrou** [comment]
@mrocklin, would you have any high-level opinion on this change?

**Comment by mrocklin** [comment]
From a performance perspective I guess that the point here is that we're making it faster to create a ProcessPoolExecutor, but making the first few tasks (or the first few times when we have enough concurrent tasks) slower.  There might be some cost in that we could have spawned many processes concurrently with greater speed than little by little during execution (maybe).

There are use cases where I would not be surprised if users wanted to create all of the processes at startup time, probably because they want more predictable performance behavior.  It might be nice if there was some convenient method to ensure that there were as many worker processes as requested workers.  The people who are likely to care about this though are probably sophisticated to map `time.sleep` or something though.

Anyway, in short I don't have strong opinions either way.  Personally, I had already assumed that the `ProcessPoolExecutor` created processes on demand (probably just because of familiarity with the `ThreadPoolExecutor`).

**Comment by aeros** [comment]
> From a performance perspective I guess that the point here is that we're making it faster to create a ProcessPoolExecutor, but making the first few tasks (or the first few times when we have enough concurrent tasks) slower.

In the issue discussion, there is also the case where a one uses the default number of max workers (`os.cpu_count()`) on a device with a large number of cores. In order for their all of their jobs to be completed concurrently, they may not need to use all of them, but end up incurring the cost to spawn all of the processes at startup, only for them to be idle for the full duration of the executor.

>  It might be nice if there was some convenient method to ensure that there were as many worker processes as requested workers. The people who are likely to care about this though are probably sophisticated to map time.sleep or something though.

It could be possible to consider an implementation of `_adjust_process_count()` that spawns a worker process for each pending work item. I.E.

```py
n_processes_needed = len(self._pending_work_items)

File: Lib/concurrent/futures/process.py [hunk 3]
```diff
@@ -601,6 +607,7 @@ def __init__(self, max_workers=None, mp_context=None,
         # Shutdown is a two-step process.
         self._shutdown_thread = False
         self._shutdown_lock = threading.Lock()
+        self._idle_worker_semaphore = mp_context.Semaphore(0)
```
Comment (pitrou) [inline Lib/concurrent/futures/process.py]: Since all semaphore operations seem to be happening in the same process, a `threading.Semaphore` should be sufficient (and lighter).

File: Lib/test/test_concurrent_futures.py [hunk 4]
```diff
@@ -964,6 +970,37 @@ def test_ressources_gced_in_workers(self):
...
+        self.assertEqual(len(executor._processes), 1)
+        executor.shutdown()
+
+    def test_idle_process_reuse_multiple(self):
+        executor = self.executor_type()
+        executor.submit(mul, 12, 7).result()
+        executor.submit(mul, 33, 25)
+        executor.submit(mul, 25, 26).result()
+        executor.submit(mul, 18, 29)
+        self.assertTrue(len(executor._processes) <= 2)
```
Comment (pitrou) [inline Lib/test/test_concurrent_futures.py]: Nit: `assertLessEqual`

File: Lib/test/test_concurrent_futures.py [hunk 5]
```diff
@@ -964,6 +970,37 @@ def test_ressources_gced_in_workers(self):
...
+
+        mp_context = get_context()
+        sem = mp_context.Semaphore(0)
+        job_count = 15 * executor._max_workers
+        for _ in range(job_count):
+            executor.submit(acquire_lock, sem)
+        self.assertEqual(len(executor._processes), executor._max_workers)
+        for _ in range(job_count):
+            sem.release()
+        executor.shutdown()
```
Comment (pitrou) [inline Lib/test/test_concurrent_futures.py]: Should use `try..finally` or, better, add `self.addCleanup(executor.shutdown)` at the beginning of this method.

File: Lib/test/test_concurrent_futures.py [hunk 6]
```diff
@@ -486,10 +486,16 @@ def _prime_executor(self):
...
-        self.executor.submit(mul, 6, 7)
-        self.executor.submit(mul, 3, 14)
-        self.assertEqual(len(self.executor._processes), 5)
+        def acquire_lock(lock):
+            lock.acquire()
+
+        mp_context = get_context()
+        sem = mp_context.Semaphore(0)
+        for _ in range(3):
+            self.executor.submit(acquire_lock, sem)
```
Comment (pitrou) [inline Lib/test/test_concurrent_futures.py:495]: A matter of taste, but this could simply be `self.executor.submit(sem.acquire)` ;-)

**Review by pitrou — COMMENTED** [review]
This looks basically good to me, just a couple comments.

File: Lib/concurrent/futures/process.py (hunk 3 above)
Comment (aeros) [reply Lib/concurrent/futures/process.py]: Agreed, good point. A previous implementation I was trying required a multiprocessing semaphore, but this version doesn't.

File: Lib/test/test_concurrent_futures.py (hunk 5 above)
Comment (aeros) [reply Lib/test/test_concurrent_futures.py]: Yeah, I would prefer `self.addCleanup(executor.shutdown)`. IMO, it helps to demonstrate that the `executor.shutdown()` isn't really part of the above test.

Edit: ended up going with a `try-finally` instead. The delayed shutdown with `self.addCleanup()` was resulting in some dangling threads (I think `threading_cleanup()` was being called prior to `executor.shutdown()`).

```
test_saturation (test.test_concurrent_futures.ProcessPoolSpawnProcessPoolExecutorTest) ... 
Warning -- threading_cleanup() failed to cleanup 2 threads (count: 2, dangling: 3)
```

Comment (aeros) [reply Lib/test/test_concurrent_futures.py:986]: Actually, I just realized something: since we can use `threading.Semaphore()`, we can simply move the existing `test_saturation` and `test_idle_thread_reuse` from `ThreadPoolExecutorTest` to `ExecutorTest`, instead of having basically the same test replicated in both `ThreadPoolExecutorTest` and `ProcessPoolExecutorTest` (including the suggested changes). I'll rename `test_idle_thread_reuse` -> `test_idle_worker_reuse`.

**Comment by aeros** [comment]
Thanks for the review @pitrou, I believe I've addressed all of the comments. Would you care to give it a final look-over? This would be my first self-authored merge.

**Review by pitrou — APPROVED** [review]
Thank you @aeros . I will merge now.

**Comment by vstinner** [comment]
@pitrou: You could have let @aeros merges the PR, he is a now a core dev 😉

**Comment by pitrou** [comment]
Ah, sorry, I hadn't thought about that.

**Comment by aeros** [comment]
No worries :)
